### PR TITLE
Add option for HTTPoison

### PR DESCRIPTION
**Pull request by derrimahendra**
This PR improves:

Add options for HTTPoison like proxy, ssl, etc

### CODE REVIEW

**Comment by binaryseed** [comment]
Hi there, can you clarify your use case for this? I don't think it's the right strategy to add this as a config option for the `new_relic_agent` app, but we can probably solve your problem another way...

**Comment by ananthakumaran** [comment]
The main usecase is the proxy config. We need to use proxy because of compliance requirement and also would like to control other options like timeout etc. 

> I don't think it's the right strategy to add this as a config option for the new_relic_agent app, but we can probably solve your problem another way

Global configuration for hackney won't solve our issue, as we don't use proxy for internal communication and we usually use different set of configs.

File: lib/new_relic/util.ex [hunk 1]
```diff
@@ -77,7 +78,9 @@ defmodule NewRelic.Util do
 
   @aws_vendor_data ["availabilityZone", "instanceId", "instanceType"]
   defp aws_vendor_hash(url) do
-    case HTTPoison.get(url, [], timeout: 100) do
+    opts = NewRelic.Config.httpoison_opts()
+
+    case HTTPoison.get(url, [], [timeout: 100] ++ opts) do
```
Comment (binaryseed) [inline lib/new_relic/util.ex]: This is making an attempt to call a port on localhost that only exists in AWS, so we wouldn't apply any proxy settings here

**Comment by ananthakumaran** [comment]
Yes, so far proxy config is the only hard requirement for us.

**Comment by binaryseed** [comment]
I'm looking into using Erlang's http client `httpc`, which can be configured to us a proxy, including a `NoProxy` setting that lets you opt-out...

http://erlang.org/doc/man/httpc.html#set_options-2

Would you be able to use that?

**Comment by ananthakumaran** [comment]
from a cursory glance, it looks like httpc provides profile feature so different set of configs can be maintained. If the agent uses the client with a specific profile, then we can configure that without affecting anything else.

**Comment by josevalim** [comment]
I would avoid httpc because it has shown multiple times it doesn't perform well under load. One option is to do something similar to ExAWS and allow this to be customizable: i.e. you can implement a function that does the request and ask the agent to use it.

@ericmj and @whatyouhide from the Elixir team are also working on a library called [xhttp](https://github.com/ericmj/xhttp) which should have v0.1 soon. The advantage is that it doesn't prescribe a process architecture, you work directly against the socket, so it can be more efficient but, on the other hand, if you need a pool, you would have to implement it yourself (but again, you can implement a pool on top of the sockets, without having multiple processes, which would be more efficient too). And of course, the library is in very early stages.

**Comment by josevalim** [comment]
@binaryseed I believe there are also concerns related to https which is tricky to set up. :(

**Comment by binaryseed** [comment]
It sure is, here's what I've done to configure https so far:

https://github.com/newrelic/elixir_agent/pull/70/files#diff-a25cfc66a1d03d0846872db375cfa09d

**Comment by ericmj** [comment]
I'm not sure that's enough, on older OTP versions I don't think the hostname will be verified against the certificate.

/cc @voltone

**Comment by binaryseed** [comment]
I could pull in `ssl_verify_fun` to do the hostname check.. I started with your code in Hex and trimmed it down as much as possible

**Comment by binaryseed** [comment]
Alternatively, I could just wait on this and circle back when `xhttp` has a stable release

**Comment by voltone** [comment]
@ericmj is right, hostname verification won't work with the options in #70. I'll leave some comments there

File: test/config_test.exs [hunk 2]
```diff
@@ -53,4 +53,16 @@ defmodule ConfigTest do
 
     assert NewRelic.Config.feature?(:error_collector)
   end
+
+  test "Can configure httpoison options" do
+    Application.put_env(:new_relic_agent, :httpoison_opts,
+      ssl: [{:versions, [:"tlsv1.2"]}],
```
Comment (voltone) [inline test/config_test.exs]: Just FYI, this is actually a dangerous example. The explicit`ssl` options override the Hackney defaults, which means peer certificate verification is not active. See below, where the first request fails (as it should) but the second one succeeds:

```elixir
iex(1)> HTTPoison.get("https://expired.badssl.com/", [], [])
{:error, %HTTPoison.Error{id: nil, reason: {:tls_alert, 'certificate expired'}}}

iex(2)> HTTPoison.get("https://expired.badssl.com/", [], ssl: [versions: [:"tlsv1.2"]])        
{:ok,
 %HTTPoison.Response{
   body: "<!DOCTYPE html>\n<html>\n<head>\n" <> ...
```

**Comment by binaryseed** [comment]
I'm closing this PR since we do not want the agent tied to a specific HTTP library. Thanks for the submission, we'll try to get your needs met another route!

**Pull request closed without merging**